neps/nep-0372.md

```diff
@@ -0,0 +1,77 @@
+---
+NEP: 372
+Title: Runtime Gas Price
+Author: Ben Kurrek <ben.kurrek@near.org>, Jakob Meier <jakob@near.org>
+DiscussionsTo: https://gov.near.org/t/expose-gas-price-and-contract-access-key-info-within-runtime/24788
+Status: Review
+Type: Standards Track
+Category: Contract
+Created: 14-Jul-2022
+---
+
+## Summary
+
+There is currently no way to query for the current Gas price at runtime from within smart contracts.
+
+## Motivation
+
+As the network becomes more popular, Gas prices may differ between blocks. By adding the ability to query for the price of Gas, it opens many possibilities for developers. Some use cases could include a contract that knows how much Gas is attached to the call or how much Gas has been used so far using the current `prepaid_gas` and `used_gas` functions but they want to convert that to a $NEAR value using the price of Gas.
+
+## Rationale and alternatives
+
+There doesn't seem to be many alternative solutions other than the one proposed. The implementation is straightforward and doesn't introduce complex behavior.
+
+## Specification
+
+Two new host functions are added to the runtime. One function is for the gas price at the current block height. The other is for the pessimistic gas price that has been used to purchase the attached gas to the currently executing receipt.
+
+Both function return a gas price measured in yoctoNEAR per unit of Gas, by writing this value as a `u128` to the guest memory address `gas_price_ptr`.
+
+```rust
+/// Gas price at the current block height.
+fn current_gas_price (gas_price_ptr: u64);
+
+/// Gas priced paid to purchase gas for the current receipt. The difference
+/// between this and the current gas price will be refunded for all the burnt
+/// gas, while unused gas will be refunded at the full pessimistic price.
+/// This function can be used to calculate how much balance is left in the
+/// account after refunds, or how much allowance will be left in an access key.
+fn pessimistic_receipt_gas_price (gas_price_ptr: u64);
+```
+
+The gas cost for both these calls will be the base host function call `wasm_base` + `wasm_write_memory_base` + 16 * `wasm_write_memory_byte`. This is in-line with similar host functions like `account_balance` that also query a single `u128` value that is readily available at the host.
+
+Both functions must fail with `MemoryAccessViolation` if [`gas_price_ptr`, `gas_price_ptr` + 16) is not completely within guest memory.
```

Comment:
Should we specify if this can be called during views or only during regular calls.

Reply:
That's a very good point. I think both functions should not be callable in views.
- `current_gas_price` would _maybe_ make sense but it would be a bit tricky to implement. Right now, view calls are oblivious of the actual as price, they just burn gas at a price of 0 NEAR/gas.
- `pessimistic_receipt_gas_price` in a view call just makes no sense, there is no receipt in a view call.